cmd/util/cmd/debug-tx/cmd.go

```diff
@@ -128,20 +193,38 @@ func run(*cobra.Command, []string) {
 		proposalKeySequenceNumber,
 	)
 
-	var txErr, processErr error
-	if flagAtLatestBlock {
-		txErr, processErr = debugger.RunTransaction(txBody)
-	} else {
-		txErr, processErr = debugger.RunTransactionAtBlockID(
-			txBody,
-			flow.Identifier(txResult.BlockID),
-			"",
-		)
-	}
-	if txErr != nil {
-		log.Fatal().Err(txErr).Msg("transaction error")
-	}
+	resultSnapshot, txErr, processErr := debugger.RunTransaction(txBody, snap, header)
 	if processErr != nil {
 		log.Fatal().Err(processErr).Msg("process error")
 	}
+
+	if flagDumpRegisters {
+		log.Info().Msg("Read registers:")
+		readRegisterIDs := resultSnapshot.ReadRegisterIDs()
+		sortRegisters(readRegisterIDs)
+		for _, registerID := range readRegisterIDs {
+			log.Info().Msgf("\t%s", registerID)
+		}
+
+		log.Info().Msg("Written registers:")
+		for _, updatedRegister := range resultSnapshot.UpdatedRegisters() {
+			log.Info().Msgf(
+				"\t%s, %s",
+				updatedRegister.Key,
+				hex.EncodeToString(updatedRegister.Value),
+			)
+		}
+	}
+	if txErr != nil {
+		log.Err(txErr).Msg("transaction error")
+	}
+}
+
+func sortRegisters(registerIDs []flow.RegisterID) {
```

Comment:
Note (for future expansion): I actually often find it most useful to have the registers sorted in read order for debugging. However currently that ordering is lost because we put the reads in a map.